components/gitpod-protocol/go/reconnecting-ws.go

```diff
@@ -48,65 +54,95 @@ func NewReconnectingWebsocket(url string, reqHeader http.Header, log *logrus.Ent
 
 // Close closes the underlying webscoket connection.
 func (rc *ReconnectingWebsocket) Close() error {
-	close(rc.closedCh)
+	rc.once.Do(func() {
+		close(rc.closedCh)
+	})
 	return nil
 }
 
-// WriteObject writes the JSON encoding of v as a message.
-// See the documentation for encoding/json Marshal for details about the conversion of Go values to JSON.
-func (rc *ReconnectingWebsocket) WriteObject(v interface{}) error {
+// EnsureConnection ensures ws connections
+// Returns only if connection is permanently failed
+func (rc *ReconnectingWebsocket) EnsureConnection(handler func(conn *WsConn) (result error, closeErr error)) error {
 	for {
-		connCh := make(chan *websocket.Conn, 1)
+		connCh := make(chan *WsConn, 1)
 		select {
 		case <-rc.closedCh:
 			return errors.New("closed")
 		case rc.connCh <- connCh:
 		}
 		conn := <-connCh
-		err := conn.WriteJSON(v)
-		if err == nil {
-			return nil
-		}
-		if !websocket.IsUnexpectedCloseError(err) {
+		err, closeErr := handler(conn)
+		if closeErr == nil {
 			return err
 		}
 		select {
 		case <-rc.closedCh:
 			return errors.New("closed")
-		case rc.errCh <- err:
+		case rc.errCh <- closeErr:
 		}
 	}
 }
 
+func isJSONError(err error) bool {
+	_, isJsonErr := err.(*json.InvalidUTF8Error)
+	if isJsonErr {
+		return true
+	}
+	_, isJsonErr = err.(*json.InvalidUnmarshalError)
+	if isJsonErr {
+		return true
+	}
+	_, isJsonErr = err.(*json.MarshalerError)
+	if isJsonErr {
+		return true
+	}
+	_, isJsonErr = err.(*json.SyntaxError)
+	if isJsonErr {
+		return true
+	}
+	_, isJsonErr = err.(*json.UnmarshalFieldError)
+	if isJsonErr {
+		return true
+	}
+	_, isJsonErr = err.(*json.UnmarshalTypeError)
+	if isJsonErr {
+		return true
+	}
+	_, isJsonErr = err.(*json.UnsupportedTypeError)
+	if isJsonErr {
+		return true
+	}
+	_, isJsonErr = err.(*json.UnsupportedValueError)
+	return isJsonErr
```

Comment:
```
[components/gitpod-protocol/go:lib] reconnecting-ws.go:118:2: S1008: should use 'return isJsonErr' instead of 'if isJsonErr { return true }; return false' (gosimple)
[components/gitpod-protocol/go:lib]     if isJsonErr {
[components/gitpod-protocol/go:lib]     ^
[components/gitpod-protocol/go:lib] package build failed
```